src/services/ckanapi.ts

```diff
@@ -0,0 +1,76 @@
+// SPDX-FileCopyrightText: 2024 PNED G.I.E.
+//
+// SPDX-License-Identifier: Apache-2.0
+import axios from 'axios';
+import { Dataset } from './../interfaces/dataset.interface';
+import { PackageSearchOptions, PackageSearchResult } from './../interfaces/packageSearch.interface';
+
+interface RawDataset {
+  [key: string]: string | string[] | null;
+}
+
+export default class CKAN {
+  DMS: string;
+
+  constructor(DMS: string) {
+    this.DMS = DMS;
+  }
+
+  async packageSearch(options: PackageSearchOptions): Promise<PackageSearchResult> {
+    const queryParams = this.constructQueryParams(options);
+    const url = `${this.DMS}/api/3/action/package_search?${queryParams}`;
+
+    try {
+      const response = await axios.get(url);
+      return {
+        datasets: this.mapDatasets(response.data.result.results),
+        count: response.data.result.count,
+      };
+    } catch (error) {
+      throw new Error(`HTTP error! ${error}`);
+    }
+  }
+
+  private buildFilterQueryPart(filters: string[]): string {
+    return filters.length > 0 ? `(${filters.join(' OR ')})` : '';
+  }
+
+  private constructQueryParams(options: PackageSearchOptions): string {
+    const groupFilter = this.buildFilterQueryPart(options.groups || []);
+    const orgFilter = this.buildFilterQueryPart(options.orgs || []);
+    const tagFilter = this.buildFilterQueryPart(options.tags || []);
+    const resFormatFilter = this.buildFilterQueryPart(options.resFormat || []);
+
+    const filters = [
+      groupFilter && `groups:${groupFilter}`,
+      orgFilter && `organization:${orgFilter}`,
+      tagFilter && `tags:${tagFilter}`,
+      resFormatFilter && `res_format:${resFormatFilter}`,
+    ]
+      .filter(Boolean)
+      .join('+');
+
+    let queryParams = `start=${options.offset || 0}&rows=${options.limit || 10}`;
+    queryParams += filters ? `&fq=${filters}` : '';
+    queryParams += options.query ? `&q=${options.query}` : '';
+    queryParams += options.sort ? `&sort=${options.sort}` : '';
+    queryParams += options.include_private ? `&include_private=${options.include_private}` : '';
+
+    return queryParams;
+  }
+
+  private toCamelCase(str: string): string {
```

Comment:
Upper case with underscore in between terms is scream snake case, not camel case.

Can you plase rename this function?

Reply:
Forget it, the code it's a bit hard to read, but now I got what it was doing, there is a first replace function that gets the first part of the string only.

Then the question is, is it really needed to convert the case?

Reply:
yes because the response is coming back in snake_case whereas javascript/typescript standard convention is camelCase.

So if we don't convert, out types will be in snake_case.

I see this code is not readable enough. I can maybe improve some parts of it to make it more understandable. Maybe can also move it into a util and with tests

Reply:
Great, thanks for the explanation. I will close this conversation and let you decide the best way to improve this code.